drivers/sensor/max17262/max17262.c

```diff
@@ -0,0 +1,345 @@
+		break;
+	case SENSOR_CHAN_GAUGE_DESIGN_VOLTAGE:
+		convert_millis(valp, config->design_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_VOLTAGE:
+		convert_millis(valp, config->desired_voltage);
+		break;
+	case SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT:
+		valp->val1 = priv->ichg_term;
+		valp->val2 = 0;
+		break;
+	case MAX17262_COULOMB_COUNTER:
+		/* Get spent capacity in mAh */
+		priv->coulomb_counter = 0xFFFF - priv->coulomb_counter;
+		valp->val1 = priv->coulomb_counter / 2;
+		valp->val2 = priv->coulomb_counter % 2 * 10 / 2;
+		break;
+	default:
+		LOG_ERR("Unsupported channel!");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Read register values for supported channels
+ *
+ * @param dev MAX17262 device to access
+ * @return 0 if successful, or negative error code from I2C API
+ */
+static int max17262_sample_fetch(const struct device *dev,
+				 enum sensor_channel chan)
+{
+	struct max17262_data *priv = dev->data;
+	struct {
+		int reg_addr;
+		int16_t *dest;
+	} regs[] = {
+		{ VCELL, &priv->voltage },
+		{ AVG_CURRENT, &priv->avg_current },
+		{ ICHG_TERM, &priv->ichg_term },
+		{ REP_SOC, &priv->state_of_charge },
+		{ INT_TEMP, &priv->internal_temp },
+		{ REP_CAP, &priv->remaining_cap },
+		{ FULL_CAP_REP, &priv->full_cap },
+		{ TTE, &priv->time_to_empty },
+		{ TTF, &priv->time_to_full },
+		{ CYCLES, &priv->cycle_count },
+		{ DESIGN_CAP, &priv->design_cap },
+		{ COULOMB_COUNTER, &priv->coulomb_counter },
+	};
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+	for (size_t i = 0; i < ARRAY_SIZE(regs); i++) {
+		int rc;
+
+		rc = max17262_reg_read(priv, regs[i].reg_addr, regs[i].dest);
+		if (rc != 0) {
+			LOG_ERR("Failed to read channel %d", chan);
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Initialise the fuel gauge
+ *
+ * @param dev MAX17262 device to access
+ * @return 0 for success
+ * @return -EINVAL if the I2C controller could not be found
+ */
+static int max17262_gauge_init(const struct device *dev)
+{
+	struct max17262_data *priv = dev->data;
+	const struct max17262_config *const config = dev->config;
+	int16_t tmp, hibcfg;
+
+	priv->i2c = device_get_binding(config->bus_name);
+	if (!priv->i2c) {
+		LOG_ERR("Could not get pointer to %s device", config->bus_name);
+		return -EINVAL;
+	}
+
+	/* Read Status register */
+	max17262_reg_read(priv, STATUS, &tmp);
+
+	if (!(tmp & STATUS_POR)) {
+		/*
+		 * Status.POR bit is set to 1 when MAX17262 detects that
+		 * a software or hardware POR event has occurred and
+		 * therefore a custom configuration needs to be set...
+		 * If POR event did not happen (Status.POR == 0), skip
+		 * init and continue with measurements.
+		 */
+		LOG_DBG("No POR event detected - skip device configuration");
+		return 0;
+	}
+	LOG_DBG("POR detected, setting custom device configuration...");
+
+	/** STEP 1 */
+	max17262_reg_read(priv, FSTAT, &tmp);
+
+	/* Do not continue until FSTAT.DNR bit is cleared */
+	while (tmp & FSTAT_DNR) {
+		k_sleep(K_MSEC(10));
```

Comment:
Why sleep at all? This is blocking the other initialization functions, so a tight loop seems like it might be a better idea, no?

Reply:
I was following the application note:
https://pdfserv.maximintegrated.com/en/an/MAX1726x-Software-Implementation-user-guide.pdf

I removed the delay for now since it works as it should without it...

Reply:
OK, thanks for clarifying. I think you should put the delay back, then. There's nothing better we can do with the current init infrastructure.

Reply:
Put the delay back.

Reply:
Agree, you should keep the delay